### DISCUSSION TITLE
Tasmota can publish data to Thingsboard?

### DISCUSSION THREAD

**Discussion by raimis3** [arendst/Tasmota#11469] [Q&A] [answered]
Is it possible for Tasmota to work with Thingsboard?

**Comment by sfromis** [comment]
The better question: Is it possible for Thingsboard to work with Tasmota?

Tasmota primarily uses MQTT messages. Can Thingsboard use MQTT publish/subscribe? On a limited level, Tasmota can also receive and send http commands. What can Thingsboard do?

**Reply by raimis3** [reply]
Hi, 
Thingsboard collect data from devices using MQTT, visualize and so on.

**Comment by barbudor** [comment, marked as answer]
From what I read here https://thingsboard.io/docs/reference/mqtt-api/
You should be able to do some things but probably not control Tasmota through it because despite Tasmota flexibility, Things board - like many of such platform - have a very strict MQTT topic format.

For example, when I read this telemetry example:
```
# Publish data as an object without timestamp (server-side timestamp will be used)
mosquitto_pub -d -q 1 -h "thingsboard.cloud" -t "v1/devices/me/telemetry" -u "$ACCESS_TOKEN" -m "{"temperature":42}"
```
You could use Tasmota `FullTopic` and `Prefix3` so it matches Thingsboard expectation.

Alternatively, you can always use Rules and `Publish` to publish on whatever topic you want.
But as you cannot disable Tasmota standard MQTT messages (such as command responses, telemetry..), I have no idea of the impact of the platform when you publish on unexpected topics.
I had similar issues with AdafruitIO

If you are looking for a full feature MQTT broker in the cloud with any constraints, you can have a look at flespi.io

**Reply by noerich777** [reply]
Hi @barbudor, 
Can you give an example how to publish ON/OFF relay status message on TASMOTA to ThingsBoard use topic "v1/devices/me/telemetry" by using Tasmota FullTopic and Prefix3 feature.

**Reply by sfromis** [reply]
For examples, see https://tasmota.github.io/docs/MQTT/
And when you use the [Publish](https://tasmota.github.io/docs/Commands/#publish) command in Tasmota, you construct the topic string yourself.

**Reply by barbudor** [reply]
Beware that connecting Tasmota to a MQTT system which enforce some topic structure, you may loose functionality.
You can change some part of the MQTT structure but not everything
Tasmota's `tele` prefix can be changed in `Telemetry`
The order of %prefix% and %topic%, adding static pieces can be done using Fulltopic
But the SENSOR message will always be published to `%fulltopic%/SENSOR`

In your exemple, `v1/devices/me/telemetry` what could be the part identifying the device (equivalent of Tasmota's %topic%) ? Is that `me` ?

**Reply by noerich777** [reply]
Hi, I get this picture from Thingsboard getting start [documentation](https://thingsboard.io/docs/getting-started-guides/helloworld/?connectdevice=mqtt-windows):

![hello-world-step-3-item-1](https://user-images.githubusercontent.com/81392444/149588921-66d24b02-81c8-45c2-ad55-eef4d6402b1d.png)
![hello-world-step-3-item-2](https://user-images.githubusercontent.com/81392444/149588924-5a0c6023-ffad-4b4e-bccc-6e70d690cf5a.png)

I have tried above step using MQTTBOX plugin for Chrome and it's work (see above pictures), ThingsBoard can subscribe this topic and telemetry tab on the specific device  shows the updated key & value, as seen as this below pictures (also taken from Thingsboard [Documentation](https://thingsboard.io/docs/getting-started-guides/helloworld/?connectdevice=mqtt-windows) :

![hello-world-step-2-item-1](https://user-images.githubusercontent.com/81392444/149589716-e5494baa-3ce5-4cae-80fa-61ae2c24562f.png)
![hello-world-step-3-item-3](https://user-images.githubusercontent.com/81392444/149589733-939c9e7a-9bca-42ce-ac2c-1a23a6ffeed3.png)

I am still find out how to do similar setting on above MQTTBOX to apply in Tasmota and send the Sonoff Switch/Toggle status continuously in some period of time (ex. Every 1 minute). My plan is to monitoring Sonoff Switch usage use ThingsBoard platform. 

I have tried configure MQTT Parameter in Tasmota then I switch ON then OFF the Sonoff Switch and the MQTT result is like this:

![image](https://user-images.githubusercontent.com/81392444/149596864-225be165-38d8-4691-a326-a26f43761f9d.png)

This topic v1/devices/me/telemetry/RESULT not recognized by ThingsBoard, and no data read on it.

Any suggestion how to make topic v1/devices/me/telemetry/RESULT change to topic **v1/devices/me/telemetry (without "RESULT")** when I switched the Sonoff to ON or OFF?

And  how to setup continuous periodically (ex. every 1 minute) send MQT: v1/devices/me/telemetry/RESULT = {"POWER1":"ON"} or {"POWER1":"OFF"}?

Thanks
_Note: I am not programmer... :)_

**Reply by barbudor** [reply]
You can reorganize part of the structure of Tasmota MQTT using Fulltopic but you full topic must always include both %topic% which represent something that is unique to this device (if you have multiple devices this is what will be different from one to another) and %prefix% which define the type of the message (will be replaced by `cmnd` for commands, `tele` by telemetry and `state` for status messages. Although you can also rename `cmnd`, `tele` and `state`to your liking.

> Any suggestion how to make topic v1/devices/me/telemetry/RESULT change to topic v1/devices/me/telemetry

The 1st approach would be to set FullTopic to `v1/devices/%topic%/%prefix%/` with Topic set to your device id (here `me`) and `tele` renamed to `telemetry` with command `Prefix3 telemetry` but that will not get you read of RESULT or SENSOR that are part of Tasmota structure.

The only viable approach is to use rules and command `Publish v1/..../telemetry  somedate` to manually publish to ThingsBoard

**Reply by noerich777** [reply]
@barbudor thank you for your explanation, I will try it.
And I am on communication with thingsboard to check can they accept sub-topic /RESULT or /SENSOR after their standard topic v1/devices/me/telemetry.

**Reply by raimis3** [reply]
@noerich777 
Hi,
Have you managed to integrate Tasmota with Thinkboard?

**Reply by noerich777** [reply]
Yes, it works!

**Reply by raimis3** [reply]
Hi, 
Fine! Thank You for answer. 
I'm trying to do according to your recommendations, but I can't get the result "latest telemetry"... maybe you can give me a secret... :)
![Untitled](https://user-images.githubusercontent.com/28362227/219764041-3cdd9e96-aacc-42d2-b08a-611cb8cba391.png)

**Reply by sfromis** [reply]
I'm not using Thingsboard, but with MQTT, the general structure should be that you make a subscription to select what MQTT topic you want to get data from, by default not anything.

**Reply by barbudor** [reply]
On mostly all commercial platform (which are more than just a cloud broker) you must publish following a specified topic structure. The platform will subscribe to random topics.

As Tasmota enforce it's own structure, it's not obvious to connect them together.
Same with AWS IOT which need special handling in Tasmota
